### PR TITLE
rocksdb: add supported_compression function

### PR DESCRIPTION
**Pull request by choleraehyq**
PTAL @andelf @zhangjinpeng1987 

I met some problems about ffi and fixed them, but I'm quite new to rust and don't know whether what I do is correct. 

Signed-off-by: Cholerae Hu <huyingqian@pingcap.com>

### CODE REVIEW

File: librocksdb_sys/crocksdb/c.cc [hunk 1]
```diff
@@ -2936,6 +2936,17 @@ const char* crocksdb_pinnableslice_value(const crocksdb_pinnableslice_t* v,
   return v->rep.data();
 }
 
+size_t crocksdb_get_supported_compression_number() {
+  return rocksdb::GetSupportedCompressions().size();
+}
+
+void crocksdb_get_supported_compression(unsigned char* v) {
```
Comment (zhangjinpeng1987) [inline librocksdb_sys/crocksdb/c.cc]: Should add a `len` parameter to indicate the expected length of the result, and validate in this function.

Comment (choleraehyq) [inline librocksdb_sys/crocksdb/c.cc]: Done.

File: librocksdb_sys/src/lib.rs [hunk 2]
```diff
@@ -71,8 +71,8 @@ pub enum DBEntryType {
     Other = 4,
 }
 
-#[derive(Copy, Clone, Debug, PartialEq)]
-#[repr(C)]
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd)]
+#[repr(u8)]
```
Comment (andelf) [inline librocksdb_sys/src/lib.rs]: Personally I think it's wrong.
Our tikv code base relying on sizeof<CompressionType> == sizeof<c_int>

File: src/rocksdb.rs [hunk 3]
```diff
@@ -1459,6 +1460,16 @@ impl Drop for SstFileWriter {
     }
 }
 
+pub fn supported_compression() -> Vec<DBCompressionType> {
+    unsafe {
+        let size = crocksdb_ffi::crocksdb_get_supported_compression_number() as usize;
+        let mut v: Vec<DBCompressionType> = Vec::with_capacity(size);
+        let pv = v.as_mut_ptr();
+        crocksdb_ffi::crocksdb_get_supported_compression(pv, size as size_t);
+        v
```
Comment (andelf) [inline src/rocksdb.rs]: Without set_len()
Vec will always be empty

File: librocksdb_sys/src/lib.rs [hunk 4]
```diff
@@ -82,6 +82,9 @@ pub enum DBCompressionType {
     DBLz4hc = 5,
     // DBXpress = 6, not support currently.
     DBZstd = 7,
+
+    // used for test
+    DBZstdNotFinal = 0x40,
```
Comment (andelf) [inline librocksdb_sys/src/lib.rs]: Do not add this

Comment (andelf) [inline librocksdb_sys/src/lib.rs]: #[cfg(test)]

Is acceptable

File: src/rocksdb.rs [hunk 5]
```diff
@@ -1889,4 +1900,12 @@ mod test {
             .unwrap();
         assert!(total_sst_files_size > 0);
     }
+
+    #[test]
+    fn test_supported_compression() {
+        let com = supported_compression();
+        for c in com {
+            assert!(c <= DBCompressionType::DBZstdNotFinal);
+        }
```
Comment (andelf) [inline src/rocksdb.rs]: Assure vec is not empty

File: librocksdb_sys/src/lib.rs [hunk 6]
```diff
@@ -82,6 +82,8 @@ pub enum DBCompressionType {
     DBLz4hc = 5,
     // DBXpress = 6, not support currently.
     DBZstd = 7,
+
+    DBZstdNotFinal = 0x40,
```
Comment (andelf) [inline librocksdb_sys/src/lib.rs]: This enum variant is useless, better remove. Since unmatched enum variant panics, ok to skip range check.

**Comment by choleraehyq** [comment]
@zhangjinpeng1987 PTAL

**Comment by siddontang** [comment]
PTAL @andelf

**Comment by andelf** [comment]
LGTM

**Comment by andelf** [comment]
Ci fails, use contains check maybe better. Since comparing CompressionType is meaningless

**Pull request merged**